File: algorithm/python/prob100_easy_Same-Tree.py
# Definition for a binary tree node.

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __hash__(self):
        MOD = 1e9+7

class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:

File: algorithm/cpp/prob1-500/prob100_easy_Same-Tree.cpp
// prob100: Same Tree

/*
 * Given two binary trees, write a function to check if they are the same or not.
 * Two binary trees are considered the same if they are structurally identical and the nodes have the same value.
 */

#include "include/Node.h"

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        return _preOrder(p, q);
    }

private:
    bool _preOrder(TreeNode* p, TreeNode* q)
    {
        if(!p && !q)
            return true;
        if(!p || !q)
            return false;
        bool result = true;
        result = result && (p -> val == q -> val);
        if(!result) return false;
        result = result && _preOrder(p -> left, q -> left);
        if(!result) return false;
        result = result && _preOrder(p -> right, q -> right);
        return result;
    }
};
